add warning if DataLoader is going to create excessive number of worker

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46867 add warning if DataLoader is going to create excessive number of worker**

This PR is to add a soft warning message to data loader when too many workers are scheduled to create.
The rule I applied here is that the total number of worker process can not exceed the total number of logical
cpus that available in the system. On Linux system, the code relies on `os.sched_getaffinity` which respects the `cpuset`. When `os.sched_getaffinity` is not available (eg. OSX, Windows), `os.cpu_count()` is used instead which does not respect the `cpuset`. (`psutil` can be used for Windows to respect `cpuset`, but I did not add this support since `psutil` is 3rd party and not always available plus the purpose of this pr is to soft warn user pay attention on worker creation when worker number is too high, not a hard stop warning).

Threading in worker process is not considered since they are currently single threaded. Each worker is set to single threaded by `torch.set_num_threads(1)`, this works well with only one exception at the moment which is QNNPack,  sometimes it doesn't `respect set_num_threads`, but given that current dataloader is not used for Mobile, this shouldn't be a problem. 
Other than `set_num_threads`, I tracked the worker process and main process see whether non-pytorch code create extra threads. The main process does have extra threads, eg. Queues, pin memory, etc. But worker thread so far is single threaded.

Differential Revision: [D24545540](https://our.internmc.facebook.com/intern/diff/D24545540)

Commit:
add warning if DataLoader might create excessive number of thread

[ghstack-poisoned]

Commit:
Update on "add warning if DataLoader might create excessive number of thread"


Differential Revision: [D24545540](https://our.internmc.facebook.com/intern/diff/D24545540)

[ghstack-poisoned]

Commit:
Update on "[WIP] add warning if DataLoader might create excessive number of thread"


Differential Revision: [D24545540](https://our.internmc.facebook.com/intern/diff/D24545540)

[ghstack-poisoned]

Commit:
Update on "[WIP] add warning if DataLoader might create excessive number of thread"


Differential Revision: [D24545540](https://our.internmc.facebook.com/intern/diff/D24545540)

[ghstack-poisoned]

ag14774 commented:
Hey, I am not sure if this check actually makes sense. If you go to the worker loop here: https://github.com/pytorch/pytorch/blob/gh/glaringlee/34/head/torch/utils/data/_utils/worker.py#L136

you can see that each worker changes the num of threads. So when you do `torch.get_num_threads()` it doesn't really reflect the actual number of threads used by each worker right?

glaringlee commented:
> Hey, I am not sure if this check actually makes sense. If you go to the worker loop here: https://github.com/pytorch/pytorch/blob/gh/glaringlee/34/head/torch/utils/data/_utils/worker.py#L136
> 
> you can see that each worker changes the num of threads. So when you do `torch.get_num_threads()` it doesn't really reflect the actual number of threads used by each worker right?

Hi, thanks for pointing this out. You are right, I noticed the (set_num_threads line), this is something I will discuss with internal people see how we deal with that line, that why this PR is [WIP].

ejguan commented on torch/utils/data/dataloader.py:
```diff
@@ -399,6 +403,90 @@ def __len__(self) -> int:
+        if not self.num_workers or self.num_workers == 0:
+            return
+
+        max_num_thread_per_worker = torch.get_num_threads()
+        max_total_thread_created = max_num_thread_per_worker * self.num_workers
```
I think `get_num_threads` is implemented here.
https://github.com/pytorch/pytorch/blob/b7e044f0e58624676130b2ef6a376a453fc6b090/aten/src/ATen/ParallelNative.cpp#L216-L234
It gives you different result from your expectation. In the `DataLoader`, it may not provide the maximum thread per worker.

glaringlee replied:
Yep, the implementation has multiple paths which result in multiple possibilities, I think normally, this api should work fine, since the main process and sub process normally has same env. But my problem currently is that whether we should check thread numbers given that we do limit thread number to 1 in the worker. I will dig this a bit, probably, we just need to suggest a worker number instead of taken threading into account.

Commit:
Update on "[WIP] add warning if DataLoader might create excessive number of thread"


Differential Revision: [D24545540](https://our.internmc.facebook.com/intern/diff/D24545540)

[ghstack-poisoned]

Commit:
Update on "add warning if DataLoader is going to create excessive number of worker"


This PR is to add a soft warning message to data loader when too many workers are scheduled to create.
The rule I applied here is that the total number of worker process can not exceed the total number of logical
cpus that available in the system. On Linux system, the code relies on `os.sched_getaffinity` which respects the `cpuset`. When `os.sched_getaffinity` is not available (eg. OSX, Windows), `os.cpu_count()` is used instead which does not respect the `cpuset`. (`psutil` can be used for Windows to respect `cpuset`, but I did not add this support since `psutil` is 3rd party and not always available plus the purpose of this pr is to soft warn user pay attention on worker creation when worker number is too high, not a hard stop warning).

Threading in worker process is not considered since they are currently single threaded. Each worker is set to single threaded by `torch.set_num_threads(1)`, this works well with only one exception at the moment which is QNNPack,  sometimes it doesn't `respect set_num_threads`, but given that current dataloader is not used for Mobile, this shouldn't be a problem. 
Other than `set_num_threads`, I tracked the worker process and main process see whether non-pytorch code create extra threads. The main process does have extra threads, eg. Queues, pin memory, etc. But worker thread so far is single threaded.

Differential Revision: [D24545540](https://our.internmc.facebook.com/intern/diff/D24545540)

[ghstack-poisoned]

VitalyFedyunin commented:
Please add another exception, as 3rd party libraries (for example numpy) doesn't respect set_num_threads too.

Commit:
Update on "add warning if DataLoader is going to create excessive number of worker"


This PR is to add a soft warning message to data loader when too many workers are scheduled to create.
The rule I applied here is that the total number of worker process can not exceed the total number of logical
cpus that available in the system. On Linux system, the code relies on `os.sched_getaffinity` which respects the `cpuset`. When `os.sched_getaffinity` is not available (eg. OSX, Windows), `os.cpu_count()` is used instead which does not respect the `cpuset`. (`psutil` can be used for Windows to respect `cpuset`, but I did not add this support since `psutil` is 3rd party and not always available plus the purpose of this pr is to soft warn user pay attention on worker creation when worker number is too high, not a hard stop warning).

Threading in worker process is not considered since they are currently single threaded. Each worker is set to single threaded by `torch.set_num_threads(1)`, this works well with only one exception at the moment which is QNNPack,  sometimes it doesn't `respect set_num_threads`, but given that current dataloader is not used for Mobile, this shouldn't be a problem. 
Other than `set_num_threads`, I tracked the worker process and main process see whether non-pytorch code create extra threads. The main process does have extra threads, eg. Queues, pin memory, etc. But worker thread so far is single threaded.

Differential Revision: [D24545540](https://our.internmc.facebook.com/intern/diff/D24545540)

[ghstack-poisoned]

Commit:
Update on "add warning if DataLoader is going to create excessive number of worker"


This PR is to add a soft warning message to data loader when too many workers are scheduled to create.
The rule I applied here is that the total number of worker process can not exceed the total number of logical
cpus that available in the system. On Linux system, the code relies on `os.sched_getaffinity` which respects the `cpuset`. When `os.sched_getaffinity` is not available (eg. OSX, Windows), `os.cpu_count()` is used instead which does not respect the `cpuset`. (`psutil` can be used for Windows to respect `cpuset`, but I did not add this support since `psutil` is 3rd party and not always available plus the purpose of this pr is to soft warn user pay attention on worker creation when worker number is too high, not a hard stop warning).

Threading in worker process is not considered since they are currently single threaded. Each worker is set to single threaded by `torch.set_num_threads(1)`, this works well with only one exception at the moment which is QNNPack,  sometimes it doesn't `respect set_num_threads`, but given that current dataloader is not used for Mobile, this shouldn't be a problem. 
Other than `set_num_threads`, I tracked the worker process and main process see whether non-pytorch code create extra threads. The main process does have extra threads, eg. Queues, pin memory, etc. But worker thread so far is single threaded.

Differential Revision: [D24545540](https://our.internmc.facebook.com/intern/diff/D24545540)

[ghstack-poisoned]

Commit:
Update on "add warning if DataLoader is going to create excessive number of worker"


This PR is to add a soft warning message to data loader when too many workers are scheduled to create.
The rule I applied here is that the total number of worker process can not exceed the total number of logical
cpus that available in the system. On Linux system, the code relies on `os.sched_getaffinity` which respects the `cpuset`. When `os.sched_getaffinity` is not available (eg. OSX, Windows), `os.cpu_count()` is used instead which does not respect the `cpuset`. (`psutil` can be used for Windows to respect `cpuset`, but I did not add this support since `psutil` is 3rd party and not always available plus the purpose of this pr is to soft warn user pay attention on worker creation when worker number is too high, not a hard stop warning).

Threading in worker process is not considered since they are currently single threaded. Each worker is set to single threaded by `torch.set_num_threads(1)`, this works well with only one exception at the moment which is QNNPack,  sometimes it doesn't `respect set_num_threads`, but given that current dataloader is not used for Mobile, this shouldn't be a problem. 
Other than `set_num_threads`, I tracked the worker process and main process see whether non-pytorch code create extra threads. The main process does have extra threads, eg. Queues, pin memory, etc. But worker thread so far is single threaded.

Differential Revision: [D24545540](https://our.internmc.facebook.com/intern/diff/D24545540)

[ghstack-poisoned]
